Custom marshaller tutorial (#31032)
dotnet/docs: docs/standard/native-interop/tutorial-custom-marshaller.md

* [`NativeMarshallingAttribute`][api_nativemarshallingattribute] &ndash; Used to indicate what marshaller to use for the attributed type. This is useful for the library authors that provide types and accompanying marshallers for those types.

These attributes, however are not the only mechanisms available to the custom marshaller author, the source generator insepct the marshaller itself for various other indications that inform how marshalling should occur.

Complete details on the design can be found in the [dotnet/runtime][design_libraryimport] repository.

### Source generator Analyzer and Fixer

Along with the source generator itself, an Analyzer and Fixer are both provided. The Analyzer and Fixer are enabled and available by default since .NET 7 RC1. The Analyzer is designed to help guide developers into using the source generator properly and the Fixer provides automated conversions from many `DllImport` patterns into the appropriate `LibraryImport` signature.

## Introducing the native library

> Using the `LibraryImport` source generator would mean consuming a native, or unmanaged, library. A native library might be a shared library (that is, `.dll`, `.so`, or `dylib`) that directly calls an operating system API, not exposed through .NET, or one that is heavily optimized in an unmanaged language that a .NET developer wants to consume. For this tutorial we are going to build our own shared library that exposes a C style API surface. The below represents a user defined type and two APIs that we would like to consume from C#. These two APIs represent the "in" mode, but there are additional modes to explore in the sample.

Reviewer: ```suggestion
Using the `LibraryImport` source generator would mean consuming a native, or unmanaged, library. A native library might be a shared library (that is, `.dll`, `.so`, or `dylib`) that directly calls an operating system API that's not exposed through .NET. The library might also be one that is heavily optimized in an unmanaged language that a .NET developer wants to consume. For this tutorial, you'll build your own shared library that exposes a C-style API surface. The following code represents a user-defined type and two APIs that you'll consume from C#. These two APIs represent the "in" mode, but there are additional modes to explore in the sample.
```

```cpp
struct error_data
{
    int code;
    bool is_fatal_error;
    char32_t* message;    /* UTF-32 encoded string */
};

extern "C" DLL_EXPORT void STDMETHODCALLTYPE PrintString(char32_t* chars);

extern "C" DLL_EXPORT void STDMETHODCALLTYPE PrintErrorData(error_data data);
```

> The above contains the two types of interest, `char32_t*` and `error_data`. The latter represents a string that is encoded in UTF-32, not an string encoding .NET historically marshals, and a user defined type containing a 32-bit integer field, a C++ boolean field, and a UTF-32 encoded string field. Both of these types will require us to provide a way for the source generator to generate marshalling code.

Reviewer: ```suggestion
The preceding code contains the two types of interest, `char32_t*` and `error_data`. `char32_t*` represents a string that's encoded in UTF-32, which is not a string encoding that .NET historically marshals. `error_data` is a user-defined type that contains a 32-bit integer field, a C++ Boolean field, and a UTF-32 encoded string field. Both of these types require you to provide a way for the source generator to generate marshalling code.
```

## Customize marshalling for a built-in type

> Let's look at the `char32_t*` type first, since marshalling this is required by the user-defined type. We should note that `char32_t*` represents the native side, but we also need representation in managed code. In .NET, there is only one "string" type, `string`. Therefore we will be marshalling a native UTF-32 encoded string to and from the `string` type in managed code. There are already several built-in marshallers for the `string` type that will marshal as UTF-8, UTF-16, ANSI, and even as the Windows `BSTR` type. However, there isn't one for marshalling as UTF-32. That is what we need to define.

Reviewer: ```suggestion
Consider the `char32_t*` type first, since marshalling this type is required by the user-defined type. `char32_t*` represents the native side, but you also need representation in managed code. In .NET, there's only one "string" type, `string`. Therefore, you'll be marshalling a native UTF-32 encoded string to and from the `string` type in managed code. There are already several built-in marshallers for the `string` type that marshal as UTF-8, UTF-16, ANSI, and even as the Windows `BSTR` type. However, there isn't one for marshalling as UTF-32. That's what you need to define.
```

> The `Utf32StringMarshaller` type has an instance of the `CustomMarshaller`, which helps to described what it does to for the source generator. Note the first type argument to the attribute is the `string` type, the managed type to marshal, the mode is next, which indicates when to use the marshaller, third type, `Utf32StringMarshaller`. Multiple instances of `CustomMarshaller` can be provided and further specify the mode and what marshaller type to use. The current example showcases a "stateless" marshaller that will take some input and return a data in the marshalled form. The `Free` method exists for symmetry with the unmanaged marshalling, the garbage collector is the "free" operation for the managed marshaller. The implementer is free to perform what operations are desired to marshal the input to the output, but remember no state will be explicitly preserved by the source generator.

Reviewer: ```suggestion
The `Utf32StringMarshaller` type has an instance of the `CustomMarshaller`, which helps to describe what it does to the source generator. The first type argument to the attribute is the `string` type, the managed type to marshal, the second is the mode, which indicates when to use the marshaller, and the third type is `Utf32StringMarshaller`. You can provide multiple instances of `CustomMarshaller` to further specify the mode and which marshaller type to use. The current example showcases a "stateless" marshaller that takes some input and returns data in the marshalled form. The `Free` method exists for symmetry with the unmanaged marshalling, and the garbage collector is the "free" operation for the managed marshaller. The implementer is free to perform whatever operations are desired to marshal the input to the output, but remember that no state will be explicitly preserved by the source generator.
```

> The specifics of how this particular marshaller performs the conversion from `string` to `char32_t*` can be found in the sample, note though that any .NET APIs could be used (for example, [`Encoding.UTF32`](https://docs.microsoft.com/dotnet/api/system.text.encoding.utf32)).

Reviewer: ```suggestion
The specifics of how this particular marshaller performs the conversion from `string` to `char32_t*` can be found in the sample. Note that any .NET APIs could be used (for example, <xref:System.Text.Encoding.UTF32?displayProperty=nameWithType>).
```

```csharp
namespace CustomMarshalling
{
    [CustomMarshaller(typeof(string), MarshalMode.Default, typeof(Utf32StringMarshaller))]
    internal static unsafe class Utf32StringMarshaller
    {
        public static uint* ConvertToUnmanaged(string? managed)
            => throw new NotImplementedException();

        public static string? ConvertToManaged(uint* unmanaged)
            => throw new NotImplementedException();

        public static void Free(uint* unmanaged)
            => throw new NotImplementedException();
    }
}
```

> Let's consider a case where state is desirable. Observe the additional `CustomMarshaller` and note the more specific mode, `MarshalMode.ManagedToUnmanagedIn`. This specialized marshaller is implemented as "stateful" and can store state across the interop call. More specialization and state permit optimizations and tailored marshalling for a mode. For example, the source generator can be instructed to provide a reasonable sized stack allocation that could avoid an explicit allocation during marshalling. The allocated buffer is requested by the marshaller when the source generator discovers a `BufferSize` property. This property indicates the amount of stack space the marshaller would like to get during the marshal call.

Reviewer: ```suggestion
Consider a case where state is desirable. Observe the additional `CustomMarshaller` and note the more specific mode, `MarshalMode.ManagedToUnmanagedIn`. This specialized marshaller is implemented as "stateful" and can store state across the interop call. More specialization and state permit optimizations and tailored marshalling for a mode. For example, the source generator can be instructed to provide a reasonably sized stack allocation that could avoid an explicit allocation during marshalling. The allocated buffer is requested by the marshaller when the source generator discovers a `BufferSize` property. This property indicates the amount of stack space the marshaller would like to get during the marshal call.
```

```csharp
namespace CustomMarshalling
{
    [CustomMarshaller(typeof(string), MarshalMode.Default, typeof(Utf32StringMarshaller))]
    [CustomMarshaller(typeof(string), MarshalMode.ManagedToUnmanagedIn, typeof(ManagedToUnmanagedIn))]
    internal static unsafe class Utf32StringMarshaller
    {
        //
        // Stateless functions removed
        //

        public ref struct ManagedToUnmanagedIn
        {
            public static int BufferSize => 0x100;

            private uint* _unmanagedValue;
            private bool _allocated; // Used stack alloc or allocated other memory

            public void FromManaged(string? managed, Span<byte> buffer)
                => throw new NotImplementedException();

            public uint* ToUnmanaged()
                => throw new NotImplementedException();

            public void Free()
                => throw new NotImplementedException();
        }
    }
}
```

> The first of our two native functions can now be called using our UTF-32 string marshallers. The below declaration uses the `LibraryImport` attribute, just like `DllImport`, but relies upon the `MarshalUsing` to instruct the source generator what marshaller to use when calling the native function. There is no need to clarify if the stateless or stateful marshaller should be used. This is handled by the implementer defining the "mode" on the marshaller. The source generator will select the most appropriate marshaller based on the context in which the `MarshalUsing` is applied, `MarshalMode.Default` being the fallback.

Reviewer: ```suggestion
You can now call the first of the two native functions using your UTF-32 string marshallers. The following declaration uses the `LibraryImport` attribute, just like `DllImport`, but relies upon the `MarshalUsing` attribute to tell the source generator which marshaller to use when calling the native function. There's no need to clarify if the stateless or stateful marshaller should be used. This is handled by the implementer defining the "mode" on the marshaller. The source generator will select the most appropriate marshaller based on the context in which the `MarshalUsing` is applied, with `MarshalMode.Default` being the fallback.
```

```csharp
// extern "C" DLL_EXPORT void STDMETHODCALLTYPE PrintString(char32_t* chars);
[LibraryImport(LibName)]
internal static partial void PrintString([MarshalUsing(typeof(Utf32StringMarshaller))] string s);
```

## Customize marshalling for a user-defined type

Marshalling a user-defined type requires defining not only the marshalling logic, but also the type in C# to marshal to/from. Let's recall the native type we are trying to marshal and define what it would ideally look like in C#.

```cpp
struct error_data
{
    int code;
    bool is_fatal_error;
    char32_t* message;    /* UTF-32 encoded string */
};
```

> An `int` is the same size in both modern C++ and in .NET. A `bool` is the canonical example for a boolean value in .NET. Building on top of `Utf32StringMarshaller`, we can marshal `char32_t*` as a .NET `string`. Accounting for .NET style the result is the following definition in C#.

Reviewer: ```suggestion
An `int` is the same size in both modern C++ and in .NET. A `bool` is the canonical example for a Boolean value in .NET. Building on top of `Utf32StringMarshaller`, you can marshal `char32_t*` as a .NET `string`. Accounting for .NET style, the result is the following definition in C#:
```

```csharp
struct ErrorData
{
    public int Code;
    public bool IsFatalError;
    public string? Message;
}
```

> Following the naming pattern, we will name the marshaller `ErrorDataMarshaller`. Instead of creating a default marshaller, we will only define marshallers for some modes. This will demonstrate the case where if a mode isn't provided, the source generator will fail. Let's start with the "in" direction.

Reviewer: ```suggestion
Following the naming pattern, name the marshaller `ErrorDataMarshaller`. Instead of creating a default marshaller, you'll only define marshallers for some modes. In this case, if a mode isn't provided, the source generator will fail. Start with defining a marshaller for the "in" direction.
```

```csharp
namespace CustomMarshalling
{
    [CustomMarshaller(typeof(ErrorData), MarshalMode.ManagedToUnmanagedIn, typeof(ErrorDataMarshaller))]
    internal static unsafe class ErrorDataMarshaller
    {
        // Unmanaged representation of ErrorData.
        // Should mimic the unmanaged error_data type at a binary level.
        internal struct ErrorDataUnmanaged
        {
            public int Code;        // .NET doesn't support less than 32-bit, so int is 32-bit.
            public byte IsFatal;    // The C++ bool is defined as a single byte.
            public uint* Message;   // This could be as simple as a void*, but uint* is closer.
        }

        public static ErrorDataUnmanaged ConvertToUnmanaged(ErrorData managed)
            => throw new NotImplementedException();

        public static void Free(ErrorDataUnmanaged unmanaged)
            => throw new NotImplementedException();
    }
}
```

> The conversion from an `ErrorData` to an `ErrorDataUnmanaged`, that will mimic the shape of the unmanaged type is now trivial with `Utf32StringMarshaller`. Note this is a "stateless" marshaller because the marshaller itself only consists of `static` functions.

Reviewer: ```suggestion
The conversion from an `ErrorData` to an `ErrorDataUnmanaged`, which will mimic the shape of the unmanaged type, is now trivial with `Utf32StringMarshaller`. This is a "stateless" marshaller because the marshaller itself only consists of `static` functions.
```

> Marshalling of an `int` is unnecessary since representations are identical in unmanaged and managed. A `bool`'s binary representation isn't defined in .NET so we use its current value to define a zero and non-zero value in the unmanaged type. We then reuse our UTF-32 marshaller to convert our `string` field into a `uint*`.

Reviewer: ```suggestion
Marshalling of an `int` is unnecessary since its representation is identical in unmanaged and managed code. A `bool` value's binary representation isn't defined in .NET, so you use its current value to define a zero and non-zero value in the unmanaged type. Then, reuse your UTF-32 marshaller to convert the `string` field into a `uint*`.
```

```csharp
public static ErrorDataUnmanaged ConvertToUnmanaged(ErrorData managed)
{
    return new ErrorDataUnmanaged
    {
        Code = managed.Code,
        IsFatal = (byte)(managed.IsFatalError ? 1 : 0),
        Message = Utf32StringMarshaller.ConvertToUnmanaged(managed.Message),
    };
}
```

> Recall we are defining this marshaller as an "in" so we must clean-up any allocations performed during the marshalling. The `int` and `bool` fields didn't allocate any memory, but the `Message` field did. We again reuse `Utf32StringMarshaller` to clean up the marshalled string.

Reviewer: ```suggestion
Recall that you're defining this marshaller as an "in", so you must clean up any allocations performed during the marshalling. The `int` and `bool` fields didn't allocate any memory, but the `Message` field did. Reuse `Utf32StringMarshaller` again to clean up the marshalled string.
```

```csharp
public static void Free(ErrorDataUnmanaged unmanaged)
    => Utf32StringMarshaller.Free(unmanaged.Message);
```

> Next, consider the case where multiple instances of `error_data` are returned. Typically, you use a collection to return multiple elements. Since this marshaller will be returning multiple elements, this is called the `MarshalMode.ElementOut` mode.

Reviewer 2: ```suggestion
Next, consider the case where multiple instances of `error_data` are returned. Typically, you use a collection to return multiple elements. The marshaller used for this scenario, corresponding to the `MarshalMode.ElementOut` mode, will be returning multiple elements.
```

> Next, we are going to consider the case where multiple instances of `error_data` are returned. Typically returning multiple of a thing is done using a collection and those multiple things are called "elements". Since this will be returning multiple elements we call this the `MarshalMode.ElementOut` mode.

Reviewer: ```suggestion
Next, consider the case where multiple instances of `error_data` are returned. Typically, you use a collection to return multiple elements. Since this marshaller will be returning multiple elements, this is called the `MarshalMode.ElementOut` mode.
```

```csharp
namespace CustomMarshalling
{
    [CustomMarshaller(typeof(ErrorData), MarshalMode.ManagedToUnmanagedIn, typeof(ErrorDataMarshaller))]
    [CustomMarshaller(typeof(ErrorData), MarshalMode.ElementOut, typeof(Out))]
    internal static unsafe class ErrorDataMarshaller
    {
        //
        // Other marshallers removed
        //

        public static class Out
        {
            public static ErrorData ConvertToManaged(ErrorDataUnmanaged unmanaged)
                => throw new NotImplementedException();

            public static void Free(ErrorDataUnmanaged unmanaged)
                => throw new NotImplementedException();
        }
    }
}
```

> The conversion from `ErrorDataUnmanaged` to `ErrorData` is the inverse of what you did for the "in" mode in the preceding code. Remember that you also need to "clean up" any allocations that the unmanaged environment expected you to perform.

Reviewer 2: ```suggestion
The conversion from `ErrorDataUnmanaged` to `ErrorData` is the inverse of what you did for the "in" mode. Remember that you also need to clean up any allocations that the unmanaged environment expected you to perform.
```

> The conversion from `ErrorDataUnmanaged` to `ErrorData` is the inverse of what you did for the "in" mode in the preceding code. Remember that you also need to "clean up" any allocations that the unmanaged environment expected you to perform.

Reviewer 2: Maybe also point out that this is stateless and only stateless marshallers are supported for `Element*` modes.

> 

Reviewer 2: This tutorial currently only shows p/invokes where the 'in' marshaller would be used. It might be helpful to also have a snippet with the p/invoke that would use this marshaller (and the one below). I think it'd also make the break/switch from talking about `ElementOut` to `ManagedToUnmanagedOut` a bit clearer.

Reviewer 2: I thnk a snippet for non-element would be helpful too. Maybe something like:

>Now, consider the case where only one `error_data` instance is returned.
>
>[c++ and p/invoke snippet]
>
>For the marshaller used in this scenario, corresponding to the `MarshalMode.ManagedToUnmanagedOut` mode, you're going to do something special...

> The conversion from `ErrorDataUnmanaged` to `ErrorData` is the inverse of what we did for the "in" mode above. Remember that we also need to "clean up" any allocations the unmanaged environment expected us to perform.

Reviewer: ```suggestion
The conversion from `ErrorDataUnmanaged` to `ErrorData` is the inverse of what you did for the "in" mode in the preceding code. Remember that you also need to "clean up" any allocations that the unmanaged environment expected you to perform.
```

> For the managed to unmanaged "out" marshaller, you're going to do something special. The name of the data type you're marshalling is called `error_data` and .NET typically expresses errors as exceptions. However, some errors are more impactful than others. Errors identified as "fatal" usually indicate a catastrophic or unrecoverable error. Notice the `error_data` has a field to check if the error is fatal. You'll marshal an `error_data` into managed code, and if it's fatal, you'll throw an exception rather than just converting it into an `ErrorData` and returning it.

Reviewer 2: ```suggestion
For the managed to unmanaged "out" marshaller, you're going to do something special. The name of the data type you're marshalling is called `error_data` and .NET typically expresses errors as exceptions. Some errors are more impactful than others and errors identified as "fatal" usually indicate a catastrophic or unrecoverable error. Notice the `error_data` has a field to check if the error is fatal. You'll marshal an `error_data` into managed code, and if it's fatal, you'll throw an exception rather than just converting it into an `ErrorData` and returning it.
```

> For our managed to unmanaged "out" marshaller, we are going to do something special. The name of the data type we are marshalling is called `error_data` and .NET typically expresses errors as exceptions. However, some errors are more impactful than others. Errors identified as "fatal" usually indicate a catastrophic or unrecoverable error. Notice the `error_data` has a field to check if the error is fatal. Let's marshal an `error_data` into managed and if it is fatal, we will throw an exception rather than just converting it into an `ErrorData` and returning it.

Reviewer: ```suggestion
For the managed to unmanaged "out" marshaller, you're going to do something special. The name of the data type you're marshalling is called `error_data` and .NET typically expresses errors as exceptions. However, some errors are more impactful than others. Errors identified as "fatal" usually indicate a catastrophic or unrecoverable error. Notice the `error_data` has a field to check if the error is fatal. You'll marshal an `error_data` into managed code, and if it's fatal, you'll throw an exception rather than just converting it into an `ErrorData` and returning it.
```

```csharp
namespace CustomMarshalling
{
    [CustomMarshaller(typeof(ErrorData), MarshalMode.ManagedToUnmanagedIn, typeof(ErrorDataMarshaller))]
    [CustomMarshaller(typeof(ErrorData), MarshalMode.ElementOut, typeof(Out))]
    [CustomMarshaller(typeof(ErrorData), MarshalMode.ManagedToUnmanagedOut, typeof(ThrowOnFatalErrorOut))]
    internal static unsafe class ErrorDataMarshaller
    {
        //
        // Other marshallers removed
        //

        public static class ThrowOnFatalErrorOut
        {
            public static ErrorData ConvertToManaged(ErrorDataUnmanaged unmanaged)
                => throw new NotImplementedException();

            public static void Free(ErrorDataUnmanaged unmanaged)
                => throw new NotImplementedException();
        }
    }
}
```

> An "out" parameter converts from an unmanaged context into a managed context, so you implement the `ConvertToManaged` method. When the unmanaged callee returns and provides an `ErrorDataUnmanaged` object, you can inspect it using your `ElementOut` mode marshaller and then check if it's marked as a "fatal" error. If so, that's your indication to throw as opposed to just returning the `ErrorData`.

Reviewer 2: ```suggestion
An "out" parameter converts from an unmanaged context into a managed context, so you implement the `ConvertToManaged` method. When the unmanaged callee returns and provides an `ErrorDataUnmanaged` object, you can inspect it using your `ElementOut` mode marshaller and check if it's marked as a fatal error. If so, that's your indication to throw instead of just returning the `ErrorData`.
```

> An "out" parameter is converting from an unmanaged context into a managed context, so we implement the `ConvertToManaged` method. When the unmanaged callee returns and provides a `ErrorDataUnmanaged`, we inspect it using our `ElementOut` mode marshaller and then check if it is marked as a "fatal" error. If so, that is our indication to throw as opposed to just return the `ErrorData`.

Reviewer: ```suggestion
An "out" parameter converts from an unmanaged context into a managed context, so you implement the `ConvertToManaged` method. When the unmanaged callee returns and provides an `ErrorDataUnmanaged` object, you can inspect it using your `ElementOut` mode marshaller and then check if it's marked as a "fatal" error. If so, that's your indication to throw as opposed to just returning the `ErrorData`.
```

```csharp
public static ErrorData ConvertToManaged(ErrorDataUnmanaged unmanaged)
{
    ErrorData data = Out.ConvertToManaged(unmanaged);
    if (data.IsFatalError)
        throw new ExternalException(data.Message, data.Code);

    return data;
}
```

> Let us assume we are not only going to consume the native library, but we are also going to share our work with the community and provide an interop library. This means we could provide `ErrorData` an implied marshaller whenever it is used in a P/Invoke by adding `[NativeMarshalling(typeof(ErrorDataMarshaller))]` to the `ErrorData` definition. With this added to `ErrorData` anyone using our definition of this type in a `LibraryImport` call will get the benefit of our marshallers. Note they can always override our marshallers by using the `MarshalUsing` at the use site.

Reviewer: ```suggestion
Perhaps you're not only going to consume the native library, but you also want to share your work with the community and provide an interop library. You can provide `ErrorData` with an implied marshaller whenever it's used in a P/Invoke by adding `[NativeMarshalling(typeof(ErrorDataMarshaller))]` to the `ErrorData` definition. Now, anyone using your definition of this type in a `LibraryImport` call will get the benefit of your marshallers. They can always override your marshallers by using `MarshalUsing` at the use site.
```

```csharp
[NativeMarshalling(typeof(ErrorDataMarshaller))]
struct ErrorData { ... }
```

>   <!-- links -->

Reviewer 2: Do we want a `See also` section with these links (and probably pinvoke-source-generation.md and custom-marshalling-source-generation.md)?

Author: These links are embedded throughout the document. I don't think they need to be called out specifically. I will cross reference `pinvoke-source-generation.md`, which seems appropriate.

> [api_custommarshallerattribute]:https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.marshalling.custommarshallerattribute

Reviewer: We usually use xref-style links for API ref links, for future reference.

---
title: Use custom marshallers in source-generated P/Invokes
description: Learn how to use the CustomMarshallerAttribute and implement a custom marshaller for use with source generation.
author: AaronRobinsonMSFT
ms.date: 08/22/2022
helpviewer_keywords:
  - "Interop"
  - "Marshalling"
  - "NativeAOT"
---

# Tutorial: Use custom marshallers in source-generated P/Invokes

In this tutorial, you'll learn how to implement a marshaller and use it for [custom marshalling](custom-marshalling-source-generation.md) in [source-generated P/Invokes](pinvoke-source-generation.md).

You'll implement marshallers for a built-in type, customize marshalling for a specific parameter and a user-defined type, and specify default marshalling for a user-defined type.

All source code used in this tutorial is available in the [dotnet/samples repository](https://github.com/dotnet/samples/tree/main/core/interop/source-generation/custom-marshalling).

## Overview of the `LibraryImport` source generator

> The [`System.Runtime.InteropServices.LibraryImportAttribute`][api_libraryimportattribute] type is the user entry point for a source generator introduced in .NET 7. This source generator is designed to generate all marshalling code at compile time instead of at run time. This has historically been done using `DllImport`, but that approach comes with costs that may not always be acceptable&mdash;see [here][pinvoke_source_generation] for more details. The `LibraryImport` source generator can generate all marshalling code and remove the run-time generation requirement intrinsic to `DllImport`.

Reviewer 2: Should we relative link or xref to these attributes these instead?

From the publishing build:
```
Line 22: [Suggestion] Absolute link 'https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.libraryimportattribute' will be broken in isolated environments. Replace with a relative link.
Line 26: [Suggestion] Absolute link 'https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.marshalling.marshalusingattribute' will be broken in isolated environments. Replace with a relative link.
Line 28: [Suggestion] Absolute link 'https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.marshalling.custommarshallerattribute' will be broken in isolated environments. Replace with a relative link.
Line 30: [Suggestion] Absolute link 'https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.marshalling.nativemarshallingattribute' will be broken in isolated environments. Replace with a relative link.
```

Author: How do I make relative links? The absolute ones are easy to find and seem stable. Constructing relative links is very tedious and non-obvious to validate.

> The [`System.Runtime.InteropServices.LibraryImportAttribute`][api_libraryimportattribute] type is the user entry point for a source generator introduced in .NET 7. This source generator is designed to generate all marshalling code at compile time instead of at run time. This has historically been done using `DllImport`, but that approach comes with costs that may not always be acceptable&mdash;see [here][pinvoke_source_generation] for more details. The `LibraryImport` source generator can generate all marshalling code and remove the run-time generation requirement intrinsic to `DllImport`.

Reviewer 2: ```suggestion
The [`System.Runtime.InteropServices.LibraryImportAttribute`][api_libraryimportattribute] type is the user entry point for a source generator introduced in .NET 7. This source generator is designed to generate all marshalling code at compile time instead of at run time. This has historically been done using `DllImport`, but that approach comes with costs that may not always be acceptable&mdash;see [P/Invoke source generation][pinvoke_source_generation] for more details. The `LibraryImport` source generator can generate all marshalling code and remove the run-time generation requirement intrinsic to `DllImport`.
```
